stdlib/public/Reflection/TypeRef.cpp

```diff
@@ -376,6 +376,297 @@ void TypeRef::dump(FILE *file, unsigned Indent) const {
   fprintf(file, "\n");
 }
 
+class DemanglingForTypeRef
+    : public TypeRefVisitor<DemanglingForTypeRef, Demangle::NodePointer> {
+  Demangle::Demangler &Dem;
+
+public:
+  DemanglingForTypeRef(Demangle::Demangler &Dem) : Dem(Dem) {}
+
+  Demangle::NodePointer visitBuiltinTypeRef(const BuiltinTypeRef *B) {
+    return Dem.demangleType(B->getMangledName());
+  }
+
+  Demangle::NodePointer visitNominalTypeRef(const NominalTypeRef *N) {
+    if (auto parent = N->getParent())
```

Comment:
The demangle tree representation for nested types is really bad because it doesn't match the language model or the textual mangling format, and every place that uses the demangle tree has to bake in workarounds for it. If you're interested, I can brain dump a refactoring that will greatly simplify this and other usages of demangle trees when nested types are involved. I'm worried that working around the bad representation will make the future refactoring more difficult by adding yet another client, and it will introduce bugs.

Reply:
> If you're interested, I can brain dump a refactoring that will greatly simplify this and other usages of demangle trees when nested types are involved.

Sure, I would at the very least like to learn what's involved, that will help decide whether it's worth the effort of doing it now ;-)